src/backend/distributed/commands/index.c

```diff
@@ -522,6 +523,80 @@ GetCreateIndexRelationLockMode(IndexStmt *createIndexStatement)
 }
 
 
+/*
+ * ReindexStmtFindRelationOid returns the oid of the relation on which the index exist
+ * if the object is an index in the reindex stmt. It returns the oid of the relation
+ * if the object is a table in the reindex stmt.
+ */
+static Oid
+ReindexStmtFindRelationOid(ReindexStmt *reindexStmt, bool missingOk)
+{
+	Assert(reindexStmt->relation != NULL);
+
+	Assert(reindexStmt->kind == REINDEX_OBJECT_INDEX ||
+		   reindexStmt->kind == REINDEX_OBJECT_TABLE);
+
+	Oid relationId = InvalidOid;
+	Relation relation = ReindexStmtFindRelation(reindexStmt, missingOk);
+
+	if (reindexStmt->kind == REINDEX_OBJECT_INDEX)
+	{
+		Oid indOid = RelationGetRelid(relation);
+		relationId = IndexGetRelation(indOid, missingOk);
+		index_close(relation, NoLock);
+	}
+	else
+	{
+		relationId = RelationGetRelid(relation);
+		table_close(relation, NoLock);
+	}
+
+	return relationId;
+}
+
+
+/*
+ * ReindexStmtFindRelation returns the relation in reindex stmt.
+ * It also acquires the relavant lock for the relation in the statement. It is caller's
+ * responsibility to close the opened relation.
+ */
+static Relation
+ReindexStmtFindRelation(ReindexStmt *reindexStmt, bool missingOk)
```

Comment:
I really don't think we need `ReindexStmtFindRelation`. The idea is to centralize the logic, now we re-distribute with multiple functions. 

 Can't the callers use:

```
Oid relOid = ReindexStmtFindRelationOid(reindexStmt, missingOk);

if (reindexStatement->kind == REINDEX_OBJECT_INDEX)
relation = index_open(indOid, NoLock);
else
relation = table_openrv(reindexStatement->relation, NoLock);

.. and do whatever you want with relation?
```